bigtable/client/internal/row_reader.h

```diff
@@ -0,0 +1,158 @@
+// Copyright 2017 Google Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef GOOGLE_CLOUD_CPP_BIGTABLE_CLIENT_INTERNAL_ROW_READER_H_
+#define GOOGLE_CLOUD_CPP_BIGTABLE_CLIENT_INTERNAL_ROW_READER_H_
+
+#include "bigtable/client/version.h"
+
+#include "bigtable/client/data_client.h"
+#include "bigtable/client/filters.h"
+#include "bigtable/client/internal/readrowsparser.h"
+#include "bigtable/client/row.h"
+#include "bigtable/client/row_set.h"
+#include "bigtable/client/rpc_backoff_policy.h"
+#include "bigtable/client/rpc_retry_policy.h"
+
+#include <google/bigtable/v2/bigtable.grpc.pb.h>
+
+#include <absl/memory/memory.h>
+#include <absl/types/optional.h>
+#include <grpc++/grpc++.h>
+
+#include <iterator>
+
+namespace bigtable {
+inline namespace BIGTABLE_CLIENT_NS {
+/**
+ * Object returned by Table::ReadRows(), enumerates rows in the response.
+ *
+ * Read rows can be obtained by iterating on this object via begin()
+ * and end().  Once an iterator is incremented, all data in the row it
+ * initially pointed to is freed.
+ *
+ */
+class RowReader {
+  class RowReaderIterator;
+
+ public:
+  /// Signifies that there is no limit on the number of rows to read.
+  static int const NO_ROWS_LIMIT = 0;
+
+  RowReader(std::shared_ptr<DataClient> client, absl::string_view table_name,
+            RowSet row_set, int rows_limit, Filter filter,
+            std::unique_ptr<RPCRetryPolicy> retry_policy,
+            std::unique_ptr<RPCBackoffPolicy> backoff_policy);
+
+  using iterator = RowReaderIterator;
+
+  /**
+   * Input iterator over rows in the response.
+   *
+   * All iterators that do not point to end() are equivalent, that is,
```

Comment:
I think we should say that creating, and particularly incrementing, multiple iterators on the sam `RowReader` is unsupported and can produce incorrect results.  We may chance our minds about the implementation later and this statement would not be true.

Consider the description for `std::istream_iterator` here:

http://en.cppreference.com/w/cpp/iterator/istream_iterator

I think we will need to say similar things: this is a single pass iterator, the data is read on increment.

Reply:
updated